aws/resource_aws_ecrpublic_repository_test.go

```diff
@@ -0,0 +1,428 @@
+package aws
+
+import (
+	"fmt"
+	"log"
+	"testing"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/ecrpublic"
+	"github.com/hashicorp/go-multierror"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/acctest"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/terraform"
+)
+
+func init() {
+	resource.AddTestSweepers("aws_ecrpublic_repository", &resource.Sweeper{
+		Name: "aws_ecrpublic_repository",
+		F:    testSweepEcrPublicRepositories,
+	})
+}
+
+func testSweepEcrPublicRepositories(region string) error {
+	client, err := sharedClientForRegion(region)
+	if err != nil {
+		return fmt.Errorf("error getting client: %w", err)
+	}
+	conn := client.(*AWSClient).ecrpublicconn
+
+	var errors error
+	err = conn.DescribeRepositoriesPages(&ecrpublic.DescribeRepositoriesInput{}, func(page *ecrpublic.DescribeRepositoriesOutput, isLast bool) bool {
+		if page == nil {
+			return !isLast
+		}
+
+		for _, repository := range page.Repositories {
+			repositoryName := aws.StringValue(repository.RepositoryName)
+			log.Printf("[INFO] Deleting ECR Public repository: %s", repositoryName)
+
+			_, err = conn.DeleteRepository(&ecrpublic.DeleteRepositoryInput{
+				// We should probably sweep repositories even if there are images.
+				Force:          aws.Bool(true),
+				RegistryId:     repository.RegistryId,
+				RepositoryName: repository.RepositoryName,
+			})
+			if err != nil {
+				if !isAWSErr(err, ecrpublic.ErrCodeRepositoryNotFoundException, "") {
+					sweeperErr := fmt.Errorf("Error deleting ECR Public repository (%s): %w", repositoryName, err)
+					log.Printf("[ERROR] %s", sweeperErr)
+					errors = multierror.Append(errors, sweeperErr)
+				}
+				continue
+			}
+		}
+
+		return !isLast
+	})
+	if err != nil {
+		if testSweepSkipSweepError(err) {
+			log.Printf("[WARN] Skipping ECR Public repository sweep for %s: %s", region, err)
+			return nil
+		}
+		errors = multierror.Append(errors, fmt.Errorf("Error retreiving ECR Public repositories: %w", err))
+	}
+
+	return errors
+}
+
+func TestAccAWSEcrPublicRepository_basic(t *testing.T) {
+	var v ecrpublic.Repository
+	rName := acctest.RandomWithPrefix("tf-acc-test")
+	resourceName := "aws_ecrpublic_repository.test"
+
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
```

Comment:
The service endpoints do not seem to be available in the usual test region `us-west-2`, however adding a `testAccPartitionHasServicePreCheck(ecrpublic.EndpointsID, t)` gives `skipping tests; partition aws does not support api.ecr-public service` even when run in a region with the service endpoints `us-east-1`.

Reply:
That `PreCheck` is based on the [AWS Go SDK v1 default endpoints information](https://github.com/aws/aws-sdk-go/blob/master/aws/endpoints/defaults.go) which is not guaranteed to be up-to-date or accurate. That information is also [very old now in v2](https://github.com/aws/aws-sdk-go-v2/blob/main/codegen/smithy-aws-go-codegen/src/main/resources/software/amazon/smithy/aws/go/codegen/endpoints.json) (if it is still used). To guard this for GovCloud and other regions, it likely will require a [Custom PreCheck](https://github.com/hashicorp/terraform-provider-aws/blob/main/docs/contributing/running-and-writing-acceptance-tests.md#custom-prechecks).